MoltenVK/MoltenVK/GPUObjects/MVKRenderPass.mm

```diff
@@ -894,6 +864,246 @@ static uint32_t getNextViewMaskGroup(uint32_t viewMask, uint32_t* startView, uin
+			attDesc.stencilStoreOp = pStencilAttInfo ? VK_ATTACHMENT_STORE_OP_STORE : VK_ATTACHMENT_STORE_OP_DONT_CARE;
+			attDesc.initialLayout = pAttInfo->resolveImageLayout;
+			attDesc.finalLayout = pAttInfo->resolveImageLayout;
+			attachmentDescriptors[attDescIdx++] = attDesc;
+		}
+	}
+}
+
+MVKRenderPass* mvkCreateRenderPass(MVKDevice* device, const VkRenderingInfo* pRenderingInfo) {
+
+	// Renderpass attachments are sequentially indexed in this order:
+	//     [color, color-resolve], ..., ds, ds-resolve
+	// skipping any attachments that do not have a VkImageView
+	uint32_t maxAttDescCnt = (pRenderingInfo->colorAttachmentCount + 1) * 2;
+	VkAttachmentDescription2 attachmentDescriptors[maxAttDescCnt];
+	VkAttachmentReference2 colorAttachmentRefs[pRenderingInfo->colorAttachmentCount];
+	VkAttachmentReference2 resolveAttachmentRefs[pRenderingInfo->colorAttachmentCount];
+
+	VkAttachmentReference2 attRef;
+	attRef.sType = VK_STRUCTURE_TYPE_ATTACHMENT_REFERENCE_2;
+	attRef.pNext = nullptr;
+	attRef.aspectMask = VK_IMAGE_ASPECT_COLOR_BIT;
+
+	uint32_t attDescIdx = 0;
+	uint32_t caRefIdx = 0;
+	bool hasClrRslvAtt = false;
+	for (uint32_t caIdx = 0; caIdx < pRenderingInfo->colorAttachmentCount; caIdx++) {
+		auto& clrAtt = pRenderingInfo->pColorAttachments[caIdx];
+		if (clrAtt.imageView) {
+			attRef.layout = clrAtt.imageLayout;
+			attRef.attachment = attDescIdx;
+			colorAttachmentRefs[caRefIdx] = attRef;
+
+			if (clrAtt.resolveImageView && clrAtt.resolveMode != VK_RESOLVE_MODE_NONE) {
+				attRef.layout = clrAtt.resolveImageLayout;
+				attRef.attachment = attDescIdx + 1;
+				resolveAttachmentRefs[caRefIdx] = attRef;
+				hasClrRslvAtt = true;
+			}
+			caRefIdx++;
+		}
+		mvkAddAttachmentDescriptor(&clrAtt, nullptr, attachmentDescriptors, attDescIdx);
+	}
+
+	// Combine depth and stencil attachments into one depth-stencil attachment.
+	// If both depth and stencil are present, their views and layouts must match.
+	VkAttachmentReference2 dsAttRef;
+	VkAttachmentReference2 dsRslvAttRef;
+	VkResolveModeFlagBits depthResolveMode = VK_RESOLVE_MODE_NONE;
+	VkResolveModeFlagBits stencilResolveMode = VK_RESOLVE_MODE_NONE;
+
+	attRef.aspectMask = 0;
+	attRef.layout = VK_IMAGE_LAYOUT_UNDEFINED;
+	VkImageLayout rslvLayout = VK_IMAGE_LAYOUT_UNDEFINED;
+
+	if (pRenderingInfo->pDepthAttachment && pRenderingInfo->pDepthAttachment->imageView) {
+		attRef.aspectMask |= VK_IMAGE_ASPECT_DEPTH_BIT;
+		depthResolveMode = pRenderingInfo->pDepthAttachment->resolveMode;
+		attRef.layout = pRenderingInfo->pDepthAttachment->imageLayout;
+		rslvLayout = pRenderingInfo->pDepthAttachment->resolveImageLayout;
+	}
+	if (pRenderingInfo->pStencilAttachment && pRenderingInfo->pStencilAttachment->imageView) {
+		attRef.aspectMask |= VK_IMAGE_ASPECT_STENCIL_BIT;
+		stencilResolveMode = pRenderingInfo->pStencilAttachment->resolveMode;
+		attRef.layout = pRenderingInfo->pStencilAttachment->imageLayout;
+		rslvLayout = pRenderingInfo->pStencilAttachment->resolveImageLayout;
+	}
```

Comment:
Metal actually supports separate depth/stencil rendering. I wonder if we couldn't implement this the other way around--that is, render passes implemented in terms of dynamic rendering.

Reply:
I did look into creating and populating some kind of mid-level objects that both dynamic rendering and fixed render passes could use. It would require moving the plumbing around populating Metal RP descriptors to those object constructs, and for now it seemed like they'd end up looking very much like the existing MVKRP object constructs that it didn't seem to be worth it, or that it would save much during dynamic behavior.

But I'm open for discussion about a v2 of all this that might optimize things better.

Reply:
I'm not sure if you're tying those changes to D/S here. But I figured we would address separate D/S as part of implementing `VK_KHR_separate_depth_stencil_layouts`, which is an extension that needs to be addressed on the road to 1.2.